contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c

```diff
@@ -1369,40 +1219,85 @@ returning_list_has_column_name(List *existing_colnames, char *current_colname)
 }
 
 /*
- * Make a function call to tsql_query_to_json_text for FOR JSON clause.
- * For example, it does the following transformation:
- * select a from t for json path =>
- * select tsql_query_to_json('select a from t', ... )
- * The first argument of the function is the query string without the for json clause.
- * The rest of the arguments passes in options and directives allowed by tsql.
- *
- * If the for json clause has TSQL variables/identifiers that needs to be binded
- * (e.g. @varName that's used as a procedure parameter) during parse analysis,
- * we transform the query to use PG's function FORMAT in order to support the
- * variable binding, for example:
- * select a from t where id = @pid for xml path =>
- * select tsql_query_to_json_text(FORMAT('select a from t where id = %s', @pid) ... )
+ * Make a function call to tsql_select_for_xml_agg() for FOR JSON clause.
  */
 ResTarget *
-TsqlForJSONMakeFuncCall(TSQL_ForClause* forclause, char* src_query, size_t start_location, core_yyscan_t yyscanner)
+TsqlForXMLMakeFuncCall(TSQL_ForClause* forclause)
+{
+	ResTarget *rt = makeNode(ResTarget);
+	FuncCall *fc;
+	List *func_name;
+	List *func_args;
+	bool binary_base64 = false;
+	bool return_xml_type = false;
+	char* root_name = NULL;
+
+	/* Resolve the XML common directive list if provided */
+	if (forclause->commonDirectives != NIL)
+	{
+		ListCell *lc;
+		foreach (lc, forclause->commonDirectives)
+		{
+			Node *myNode = lfirst(lc);
+			A_Const *myConst;
+
+			/* commonDirective is either integer const or string const */
+			Assert(IsA(myNode, A_Const));
+			myConst = (A_Const *)myNode;
+			Assert(myConst->val.type == T_Integer || myConst->val.type == T_String);
+			if (myConst->val.type == T_Integer)
+			{
+				if (myConst->val.val.ival == TSQL_XML_DIRECTIVE_BINARY_BASE64)
+					binary_base64 = true;
+				else if (myConst->val.val.ival == TSQL_XML_DIRECTIVE_TYPE)
+					return_xml_type = true;
+			}
+			else if (myConst->val.type == T_String)
+			{
+				root_name = myConst->val.val.str;
+			}
+		}
+	}
+
+	/*
+	 * Finally make function call to tsql_select_for_xml_agg or tsql_select_for_xml_text_agg
+	 * depending on the return_xml_type flag (TYPE option in the FOR XML clause).
+	 * The only difference of the two functions is the return type. tsql_select_for_xml_agg
+	 * returns XML type, tsql_select_for_xml_text_agg returns text type.
+	 */
+	if (return_xml_type)
+		func_name= list_make2(makeString("sys"), makeString("tsql_select_for_xml_agg"));
+	else
+		func_name= list_make2(makeString("sys"), makeString("tsql_select_for_xml_text_agg"));
+	func_args = list_make5(makeColumnRef(construct_unique_index_name("rows", "forjson"), NIL, -1, NULL),
```

Comment:
1. I can change the name to something else, but it does need to be the same for both `for json` and `for xml` since they share the same `TsqlForClauseSubselect` usage. I'll probably use something like `tsql_for`.
2. `construct-unique-index-name()` is overall a pretty simple function, it just does an md5 hash of the first argument and then appends the second argument to the end - the only case when the error would be raised is if the md5 itself fails, which is extremely unlikely and would basically only happen if the system hits OOM per the comments in the md5 function. I don't think it is worth creating a new function for this case, but I'll add some comments explaining why this is used in the first place.